content/architecture/storage-engine/prolly-tree.md

```diff
@@ -2,38 +2,234 @@
+
+![Building a Prolly-Tree, Step 1](../../.gitbook/assets/tim-prolly-tree-step-1.png)
+
+2. **Determine Chunk Boundaries**: Use a seed, the size of the current chunk, and the key value to calculate a rolling hash. Any time the hash value is below a target value, form a chunk boundary and start a new block. Here's the chunking step on our leaf nodes:
+
+![Building a Prolly-tree, Step 2](../../.gitbook/assets/tim-prolly-tree-step-2.png)
+
+3. **Hash each Chunk**: You now have the leaf nodes of the Prolly Tree. Compute the content address of each block by applying a strong hash function to its contents. You store the contents in a content addressed block store. Here our blocks have been addressed and stored in the block store:
+
+![Building a Prolly-tree, Step 3](../../.gitbook/assets/tim-prolly-tree-step-3.png)
+
+4. **Finished?**: If the length of your list of content addresses is 1, then you are done. This is the content address of your tree.
+
+5. **Build a New Map** Otherwise, form the next layer of your Prolly Tree by creating a map of the highest key value in the chunk and the content address of the chunk. Here's what the entries for the first internal level of the tree would look like for our example:
+
+![Building a Prolly-tree, Step 5](../../.gitbook/assets/tim-prolly-tree-step-5.png)
+
+6. **Return to Step 2**: Use this new map step 2. The algorithm terminates when step 4's condition is reached.
+
+Following these steps results in the following Prolly Tree. 
+
+![Prolly Tree](../../.gitbook/assets/tim-prolly-tree-example.png)
+
+## Modifying a Prolly Tree
+
+The magic of Prolly Trees is not seen on construction, but on modification. As you'll see, modifying a Prolly Tree changes the minimal amount of content addresses. Minimal modification means more structural sharing of contents because fewer content addresses change. It is also the basis of fast diff which is accomplished by comparing content addresses.
+
+### Update a Value
+
+If we update a value, we walk the tree by key to the leaf node holding the value. We then edit the referenced chunk in place. After the edit, we recalculate the content address of the chunk. We then walk up the tree recalculating each internal content address up to the root of the tree. 
+
+![Prolly Tree Value Update](../../.gitbook/assets/prolly-tree-update-value.png)
+
+### Insert Keys
+
+We can insert keys at the beginning, middle, or end of the key space. It's helpful to visualize each type of insert.
+
+When we insert a key, we walk the tree by key to the leaf node where the key belongs. We then edit the chunk, calculate whether to split the chunk or not, and then recalculate the content address. Note, a key insert or delete has a small probability of splitting the chunk. If the rolling hash value when incorporating the new key randomly comes up below our boundary threshold, the chunk is split at the new boundary. Finally, we walk up the tree recalculating each content address up to the root of the tree.
```

Comment:
Maybe something like:

Note, a key insert or a delete has a small probability of changing an existing chunk boundary. If the computed hash values for the keys, combined with the size of chunk at those keys, happens to chose a different chunk boundary, the chunk will be split at the new boundary.